### PR TITLE
fix: allow ca certificate requests

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Fix an issue where requests for CA certificates were not managed correctly. The CertificateRequest was made for a CA certificate, the lib would create the correct CSR, then would remove the CSR from the relation databag thinking it's not a match.

## This is a major api change
Here we get rid of the `CSR.to_certificate_request` and replace it with the `CertificateRequest.from_csr`. Since this library is still in beta, we don't go with a major version bump. 

## Checklist

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added tests that validate the behaviour of the software.
- [ ] I validated that new and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I have bumped the version of any required library.

### CODE REVIEW

File: tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4.py [hunk 1]
```diff
@@ -273,7 +273,8 @@ def test_given_ca_certificate_attributes_when_generate_ca_then_ca_is_generated_c
     assert ca_certificate.sans_oid == frozenset()
 
 
-# Generate Certificate
+# # Generate Certificate
```
Comment (saltiyazan) [inline tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4.py:276]: nitpicking, but why the change here?

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 2]
```diff
@@ -353,9 +353,10 @@ class Certificate:
 
     raw: str
     common_name: str
-    sans_dns: Optional[FrozenSet[str]] = None
-    sans_ip: Optional[FrozenSet[str]] = None
-    sans_oid: Optional[FrozenSet[str]] = None
+    is_ca: bool
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py]: Done

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 3]
```diff
@@ -412,9 +413,18 @@ def from_string(cls, certificate: str) -> "Certificate":
             sans_oid = []
         expiry_time = certificate_object.not_valid_after_utc
         validity_start_time = certificate_object.not_valid_before_utc
+        is_ca: bool = False
+        try:
+            is_ca = certificate_object.extensions.get_extension_for_oid(
+                ExtensionOID.BASIC_CONSTRAINTS
+            ).value.ca  # type: ignore[reportAttributeAccessIssue]
+        except x509.ExtensionNotFound:
+            is_ca = False
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py]: Sure, done

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 4]
```diff
@@ -412,9 +413,18 @@ def from_string(cls, certificate: str) -> "Certificate":
...
+
         return cls(
             raw=certificate.strip(),
             common_name=str(common_name[0].value),
+            is_ca=is_ca,
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:427]: We can't really go with this recommendation as is_ca would sometimes be unbound. I kept the is_ca = False before the try/except statement

File: tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4.py (hunk 1 above)
Comment (gruyaume) [reply tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4.py]: No reason, removed now.

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 5]
```diff
@@ -458,22 +467,6 @@ def __str__(self) -> str:
         """Return the CSR as a string."""
         return self.raw
 
-    def to_certificate_request(self) -> "CertificateRequest":
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:461]: I added a not in the PR description

File: lib/charms/tls_certificates_interface/v4/tls_certificates.py [hunk 6]
```diff
@@ -1269,9 +1280,14 @@ def _private_key_generated(self) -> bool:
             return False
         return True
 
-    def _csr_matches_certificate_request(self, csr: CertificateSigningRequest) -> bool:
+    def _csr_matches_certificate_request(
```
[the remark this answers is not in the document]
Comment (gruyaume) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:1283]: At the moment the is_ca field is a separate value in the charm relation interface databag. I agree that we should likely move to having it in the CSR, but this would impact the charm relation interface. There we would likely have to deprecate the field in the databag and encourage users to take the CSR route. 

This should go in a separate change.

[the remark this answers is not in the document]
Comment (saltiyazan) [reply lib/charms/tls_certificates_interface/v4/tls_certificates.py:1283]: Alright.

**Review by saltiyazan — APPROVED** [review]
